drivers/hmis/app/graphql/mutations/bulk_assign_service.rb

```diff
@@ -0,0 +1,106 @@
+    field :success, Boolean, null: true
+
+    def resolve(input:)
+      project = Hmis::Hud::Project.viewable_by(current_user).find_by(id: input.project_id)
+      access_denied! unless project
+      access_denied! unless current_permission?(permission: :can_edit_enrollments, entity: project)
+
+      clients = Hmis::Hud::Client.viewable_by(current_user).where(id: input.client_ids).preload(:enrollments)
+      access_denied! unless clients.count == input.client_ids.uniq.length
+
+      cst = Hmis::Hud::CustomServiceType.find(input.service_type_id)
+      hud_user_id = Hmis::Hud::User.from_user(current_user).user_id
+
+      # Evaluate perm, but don't raise an exception unless the operation actually needs to perform an enrollment
+      can_enroll_clients = current_permission?(permission: :can_enroll_clients, entity: project)
+
+      # Determine and validate CoC Code, which is needed for creating new Enrollments
+      coc_code = determine_coc_code(coc_code_arg: input.coc_code, project: project)
+
+      project_has_units = project.units.exists?
+      available_units = project.units.unoccupied_on(input.date_provided).order(updated_at: :desc).to_a
+
+      Hmis::Hud::Service.transaction do
+        clients.each do |client|
+          # Look for Enrollment at the project that is open on the service date
+          enrollment = load_open_enrollment_for_client(
+            client,
+            project_id: project.id,
+            open_on_date: input.date_provided,
+          )
+
+          # If no Enrollment was found, create one
+          unless enrollment
+            enrollment = Hmis::Hud::Enrollment.new(
+              client: client,
+              project: project,
+              entry_date: input.date_provided,
+              user_id: hud_user_id,
+              household_id: Hmis::Hud::Base.generate_uuid,
+              enrollment_coc: coc_code,
+              relationship_to_hoh: 1, # Head of Household
+            )
+            access_denied! unless can_enroll_clients
+            raise 'bulk service assignment generated invalid enrollment' unless enrollment.valid?
+
+            # Attempt to assign this enrollment to a unit if this project has units. This is AC-specific for now, and does
+            # not support specifying the unit type. Needs improvement if/when we expand unit capabilities.
+            if project_has_units
+              error_out('Failed to enroll client because there are no available units.') if available_units.empty?
+
+              enrollment.assign_unit(unit: available_units.pop, start_date: input.date_provided, user: current_user)
+            end
+
+            enrollment.save_new_enrollment!
+          end
+
+          # Initialize using the HmisService view. Based on the CustomServiceType, the class will initialize
+          # either a Hmis::Hud::Service or Hmis::Hud::CustomService as the `owner`
+          service = Hmis::Hud::HmisService.new(
```

Comment:
Note: I know this has been around for a while but it still strikes me as odd to use HmisService as a factory. Maybe we could revisit this in future and find a better home for this responsibility